azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -48,3 +63,331 @@ def _execute_op(self, op):
+            )
+            logger.error("response body: {}".format(provisioning_op.response_body))
+            return exceptions.ServiceError(
+                "{prov_op_name} request returned a service error status code {status_code}".format(
+                    prov_op_name=prov_op_name, status_code=provisioning_op.status_code
+                )
+            )
+        else:
+            return None
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _decode_response(self, provisioning_op):
+        return json.loads(provisioning_op.response_body.decode("utf-8"))
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _get_registration_status(self, decoded_response):
+
+        status = None if "status" not in decoded_response else str(decoded_response["status"])
+        return status
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _get_operation_id(self, decoded_response):
+
+        operation_id = (
+            None if "operationId" not in decoded_response else str(decoded_response["operationId"])
+        )
+        return operation_id
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _form_complete_result(self, operation_id, decoded_response, status):
+        """
+        Create the registration result from the complete decoded json response for details regarding the registration process.
+        """
+        decoded_state = (
+            None
+            if "registrationState" not in decoded_response
+            else decoded_response["registrationState"]
+        )
+        registration_state = None
+        if decoded_state is not None:
+            # Everything needs to be converted to string explicitly for python 2
+            # as everything is by default a unicode character
+            registration_state = RegistrationState(
+                device_id=None
+                if "deviceId" not in decoded_state
+                else str(decoded_state["deviceId"]),
+                assigned_hub=None
+                if "assignedHub" not in decoded_state
+                else str(decoded_state["assignedHub"]),
+                sub_status=None
+                if "substatus" not in decoded_state
+                else str(decoded_state["substatus"]),
+                created_date_time=None
+                if "createdDateTimeUtc" not in decoded_state
+                else str(decoded_state["createdDateTimeUtc"]),
+                last_update_date_time=None
+                if "lastUpdatedDateTimeUtc" not in decoded_state
+                else str(decoded_state["lastUpdatedDateTimeUtc"]),
+                etag=None if "etag" not in decoded_state else str(decoded_state["etag"]),
+            )
+
+        registration_result = RegistrationResult(
+            # request_id=request_id,
+            operation_id=operation_id,
+            status=status,
+            registration_state=registration_state,
+        )
+        return registration_result
+
+    @abc.abstractmethod
+    def _execute_op(self, op):
```

Comment:
You shouldn't need this.  It should inherit the abstract method from the base class.

Reply:
probably it was giving me a warning saying I must have all abstract methods...but yes its not needed...

---
In reply to: [345907439](https://github.com/Azure/azure-iot-sdk-python/pull/391#discussion_r345907439) [](ancestors = 345907439)